packages/name-controller/src/providers/etherscan.ts

```diff
@@ -0,0 +1,101 @@
+import { ETHERSCAN_SUPPORTED_NETWORKS } from '../constants';
+import type {
+  NameProvider,
+  NameProviderMetadata,
+  NameProviderRequest,
+  NameProviderResult,
+} from '../types';
+import { NameType } from '../types';
+import { handleFetch } from '../util';
+
+const ID = 'etherscan';
+const LABEL = 'Etherscan (Verified Contract Name)';
+
+type EtherscanGetSourceCodeResponse = {
+  status: '1' | '0';
+  message: string;
+  result: [
+    {
+      SourceCode: string;
+      ABI: string;
+      ContractName: string;
+      CompilerVersion: string;
+      OptimizationUsed: string;
+      Runs: string;
+      ConstructorArguments: string;
+      Library: string;
+      LicenseType: string;
+      Proxy: string;
+      Implementation: string;
+      SwarmSource: string;
+    },
+  ];
+};
+
+export class EtherscanNameProvider implements NameProvider {
+  #apiKey?: string;
+
+  constructor({ apiKey }: { apiKey?: string } = {}) {
+    this.#apiKey = apiKey;
+  }
+
+  getMetadata(): NameProviderMetadata {
+    return {
+      sourceIds: { [NameType.ETHEREUM_ADDRESS]: [ID] },
+      sourceLabels: { [ID]: LABEL },
+    };
+  }
+
+  async getProposedNames(
+    request: NameProviderRequest,
+  ): Promise<NameProviderResult> {
+    const { value, chainId } = request;
+
+    const url = this.#getUrl(chainId, {
+      module: 'contract',
+      action: 'getsourcecode',
+      address: value,
+      apikey: this.#apiKey,
+    });
+
+    const responseData = (await handleFetch(
+      url,
+    )) as EtherscanGetSourceCodeResponse;
+
+    const results = responseData?.result ?? [];
+    const proposedNames = results.map((result) => result.ContractName);
+
+    return {
+      results: {
+        [ID]: {
+          proposedNames,
+        },
+      },
+    };
+  }
+
+  #getUrl(chainId: string, params: Record<string, string | undefined>): string {
+    type SupportedChainId = keyof typeof ETHERSCAN_SUPPORTED_NETWORKS;
+
+    const networkInfo =
+      ETHERSCAN_SUPPORTED_NETWORKS[chainId as SupportedChainId];
```

Comment:
I think this would be a great solution if we were reusing the supported chain ID in many places, however currently it's used only once and immediately verified to be a valid key.

Because of that, I fear introducing a separate method would only increase the code and mean we're searching `ETHERSCAN_SUPPORTED_NETWORKS` twice rather than once.

Reply:
Gotcha. Well, I like that you're checking for the existence of `networkInfo` — that mitigates the risk that `chainId` is not a key of `ETHERSCAN_SUPPORTED_NETWORKS` already, so no problem there.